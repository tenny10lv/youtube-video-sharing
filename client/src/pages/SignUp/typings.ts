export enum Field {
  EMAIL = "email",
  PASSWORD = "password",
  CONFIRM_PASSWORD = "confirmPassword",
}

export type FormField = {
  [Field.EMAIL]: string;
  [Field.PASSWORD]: string;
  [Field.CONFIRM_PASSWORD]: string;
};
