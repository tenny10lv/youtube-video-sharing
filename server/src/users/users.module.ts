import { Module } from '@nestjs/common';
import { SequelizeModule } from '@nestjs/sequelize';
import { UsersService } from './users.service';
import { UsersController } from './users.controller';
import { User } from './models/user.model';
import { UsersProvider } from './users.provider';

@Module({
  controllers: [UsersController],
  providers: [UsersService, ...UsersProvider],
  imports: [SequelizeModule.forFeature([User])],
  exports: [SequelizeModule, UsersService],
})
export class UsersModule {}
