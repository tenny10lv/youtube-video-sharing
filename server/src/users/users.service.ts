import { Injectable } from '@nestjs/common';
import { InjectModel } from '@nestjs/sequelize';
import { CrudService } from 'nestjs-crud-sequelize';
import { User } from './models/user.model';

@Injectable()
export class UsersService extends CrudService<User> {
  constructor(
    @InjectModel(User)
    model: typeof User,
  ) {
    super(model);
  }
}
