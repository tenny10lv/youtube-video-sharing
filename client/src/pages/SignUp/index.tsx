import React, { useCallback, useState } from "react";
import { useNavigate } from "react-router-dom";
import {
  Container,
  CssBaseline,
  Box,
  Typography,
  TextField,
  Button,
} from "@mui/material";
import { useFormik } from "formik";
import * as Yup from "yup";
import { toast } from "react-toastify";
import { Field, FormField } from "./typings";
import { registerUser } from "../../api/users";

const signupSchema = Yup.object().shape({
  email: Yup.string().email("Invalid email").required("Required"),
  password: Yup.string()
    .min(6, "Password must be at least 6 characters")
    .required("Required"),
  confirmPassword: Yup.string()
    .min(6, "Confirm password must be at least 6 characters")
    .required("Required")
    .when("password", {
      is: (val: string) => (val && val.length > 0 ? true : false),
      then: Yup.string().oneOf(
        [Yup.ref("password")],
        "Both password need to be the same"
      ),
    }),
});

export default function SignUp() {
  const navigate = useNavigate();
  const [isSubmitting, setIsSubmitting] = useState(false);

  const onSubmit = async (values: FormField) => {
    try {
      setIsSubmitting(true);
      const { data } = await registerUser({
        username: values[Field.EMAIL],
        password: values[Field.PASSWORD],
      });
      if (data) {
        toast("Register successfully", { type: "success" });
      }
    } catch (error) {
      toast("Register failed", { type: "error" });
    } finally {
      setIsSubmitting(false);
    }
  };

  const formik = useFormik<FormField>({
    initialValues: {
      [Field.EMAIL]: "",
      [Field.PASSWORD]: "",
      [Field.CONFIRM_PASSWORD]: "",
    },
    validationSchema: signupSchema,
    onSubmit,
  });

  const getError = useCallback(
    (field: Field) => {
      return formik.touched[field] && formik.errors[field];
    },
    [formik]
  );

  const handleSubmit = (event: any) => {
    event.preventDefault();
    formik.handleSubmit();
  };

  const onLoginClickHandler = () => {
    navigate("/login");
  };

  return (
    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <Box
        display="flex"
        flexDirection="column"
        justifyContent="center"
        alignItems="center"
        mt={8}
      >
        <Typography component="h1" variant="h5">
          Sign Up
        </Typography>
        <Box component="form" onSubmit={handleSubmit} sx={{ mt: 1 }}>
          <TextField
            name={Field.EMAIL}
            label="Email Address *"
            fullWidth
            margin="normal"
            error={!!getError(Field.EMAIL)}
            helperText={getError(Field.EMAIL)}
            value={formik.values.email}
            onChange={formik.handleChange}
          />
          <TextField
            name={Field.PASSWORD}
            label="Password *"
            type="password"
            fullWidth
            margin="normal"
            error={!!getError(Field.PASSWORD)}
            helperText={getError(Field.PASSWORD)}
            value={formik.values.password}
            onChange={formik.handleChange}
          />
          <TextField
            name={Field.CONFIRM_PASSWORD}
            label="Confirm Password *"
            type="password"
            fullWidth
            margin="normal"
            error={!!getError(Field.CONFIRM_PASSWORD)}
            helperText={getError(Field.CONFIRM_PASSWORD)}
            value={formik.values.confirmPassword}
            onChange={formik.handleChange}
          />
          <Button
            type="submit"
            fullWidth
            variant="contained"
            disabled={isSubmitting}
            sx={{ mt: 3, mb: 3 }}
          >
            Submit
          </Button>
          <Button
            fullWidth
            variant="text"
            sx={{ mb: 2 }}
            onClick={onLoginClickHandler}
          >
            Log In
          </Button>
        </Box>
      </Box>
    </Container>
  );
}
