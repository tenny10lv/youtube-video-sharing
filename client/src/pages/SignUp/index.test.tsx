import React from "react";
import { fireEvent, render, screen } from "@testing-library/react";
import { ToastContainer } from "react-toastify";
import * as userApi from "../../api/users";
import SignUp from "./";

const mockedUsedNavigate = jest.fn();
jest.mock("react-router-dom", () => ({
  ...(jest.requireActual("react-router-dom") as any),
  useNavigate: () => mockedUsedNavigate,
}));

describe("signup-page", () => {
  test("Should match snapshot", () => {
    const { container } = render(<SignUp />);
    expect(container).toMatchSnapshot();
  });

  test("render signup page", async () => {
    render(<SignUp />);

    const emailInput = screen.getByLabelText("Email Address *") as any;
    expect(emailInput).toBeInTheDocument();

    const passwordInput = screen.getByLabelText("Password *") as any;
    expect(passwordInput).toBeInTheDocument();

    const confimPasswordInput = screen.getByLabelText(
      "Confirm Password *"
    ) as any;
    expect(confimPasswordInput).toBeInTheDocument();

    const signUpButton = screen.getByRole("button", {
      name: "Submit",
    });
    expect(signUpButton).toBeInTheDocument();
  });

  test("signup page - validate input", async () => {
    render(<SignUp />);

    const signUpButton = screen.getByRole("button", {
      name: "Submit",
    });

    // empty inputs
    fireEvent.click(signUpButton);
    const requiredTexts = await screen.findAllByText("Required");
    expect(requiredTexts.length).toEqual(3);
  });

  test("sign up page - sign up successfully", async () => {
    jest.spyOn(userApi, "registerUser").mockResolvedValue({ data: {} } as any);

    render(
      <>
        <ToastContainer autoClose={3000} position="bottom-right" />
        <SignUp />
      </>
    );

    const emailInput = screen.getByLabelText("Email Address *") as any;
    fireEvent.change(emailInput, {
      target: { value: "test@yopmail.com" },
    });

    const passwordInput = screen.getByLabelText("Password *") as any;
    fireEvent.change(passwordInput, {
      target: { value: "123456" },
    });

    const cfPasswordInput = screen.getByLabelText("Confirm Password *") as any;
    fireEvent.change(cfPasswordInput, {
      target: { value: "123456" },
    });

    const submitButton = screen.getByRole("button", {
      name: "Submit",
    });
    fireEvent.click(submitButton);

    expect(await screen.findByText("Register successfully")).toBeVisible();
  });

  test("sign up page - sign up failed", async () => {
    jest.spyOn(userApi, "registerUser").mockRejectedValue({});

    render(
      <>
        <ToastContainer autoClose={3000} position="bottom-right" />
        <SignUp />
      </>
    );

    const emailInput = screen.getByLabelText("Email Address *") as any;
    fireEvent.change(emailInput, {
      target: { value: "test@yopmail.com" },
    });

    const passwordInput = screen.getByLabelText("Password *") as any;
    fireEvent.change(passwordInput, {
      target: { value: "123456" },
    });

    const cfPasswordInput = screen.getByLabelText("Confirm Password *") as any;
    fireEvent.change(cfPasswordInput, {
      target: { value: "123456" },
    });

    const submitButton = screen.getByRole("button", {
      name: "Submit",
    });
    fireEvent.click(submitButton);

    expect(await screen.findByText("Register failed")).toBeVisible();
  });

  test("sign up page - go to login", async () => {
    render(<SignUp />);

    const loginButton = screen.getByRole("button", {
      name: "Log In",
    });
    fireEvent.click(loginButton);

    expect(mockedUsedNavigate).toHaveBeenCalledWith("/login");
  });
});
