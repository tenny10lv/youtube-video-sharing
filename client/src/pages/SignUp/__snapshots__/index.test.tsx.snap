// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`signup-page Should match snapshot 1`] = `
<div>
  <main
    class="MuiContainer-root MuiContainer-maxWidthXs css-nzinn5-MuiContainer-root"
  >
    <div
      class="MuiBox-root css-1oitm5x"
    >
      <h1
        class="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
      >
        Sign Up
      </h1>
      <form
        class="MuiBox-root css-164r41r"
      >
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r0:"
            id=":r0:-label"
          >
            Email Address *
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r0:"
              name="email"
              type="text"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Email Address *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r1:"
            id=":r1:-label"
          >
            Password *
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              name="password"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Password *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          class="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth MuiTextField-root css-17vbkzs-MuiFormControl-root-MuiTextField-root"
        >
          <label
            class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink="false"
            for=":r2:"
            id=":r2:-label"
          >
            Confirm Password *
          </label>
          <div
            class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
          >
            <input
              aria-invalid="false"
              class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r2:"
              name="confirmPassword"
              type="password"
              value=""
            />
            <fieldset
              aria-hidden="true"
              class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                class="css-1ftyaf0"
              >
                <span>
                  Confirm Password *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <button
          class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-masstu-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="submit"
        >
          Submit
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-fullWidth MuiButtonBase-root  css-s0sliy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Log In
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </form>
    </div>
  </main>
</div>
`;
